integration/spring-boot/autoconfigure/src/main/java/org/apache/cxf/spring/boot/autoconfigure/micrometer/MicrometerObservationAutoConfiguration.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cxf.spring.boot.autoconfigure.micrometer;
+
+import org.apache.cxf.tracing.micrometer.MessageInObservationConvention;
+import org.apache.cxf.tracing.micrometer.MessageOutObservationConvention;
+import org.apache.cxf.tracing.micrometer.ObservationClientFeature;
+import org.apache.cxf.tracing.micrometer.ObservationFeature;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.actuate.autoconfigure.observation.ObservationAutoConfiguration;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import io.micrometer.observation.ObservationRegistry;
+
+@Configuration
+@AutoConfigureAfter(ObservationAutoConfiguration.class)
+@ConditionalOnClass({ ObservationRegistry.class, ObservationFeature.class })
+@ConditionalOnProperty(name = "cxf.observation.enabled", matchIfMissing = true)
+@ConditionalOnBean(ObservationRegistry.class)
+public class MicrometerObservationAutoConfiguration {
```

Comment:
I am unsure about this particular auto configuration, it conflicts with the way CXF uses scans (coming from CxfAutoConfiguration).

Reply:
The `MessageInObservationConvention` and `MessageOutObservationConvention` *MAY BE* provided by the user. This is to override the defaults that CXF would provide. It is perfectly valid *NOT* to have these defined by the user

Reply:
So `CxfAutoConfiguration` will create a bean for the `ObservationFeature` ? I can't find the code that does that.

Reply:
Yes, there are 2 way it works (managed by property):
 -  scanning classpath and extracting JAX-RS annotated classes (done by https://github.com/apache/cxf/blob/main/rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/spring/AbstractJaxrsClassesScanServer.java)
 - scanning Spring components (done by https://github.com/apache/cxf/blob/main/rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/spring/AbstractSpringComponentScanServer.java)

Reply:
Correct, that is why I was suggesting to have pre-created all needed beans (guarded by @ConditionalOnMissingBean, `ObservationRegistry` falls into this category) in this auto configuration (if there are chances in may not be present. But you think it is redundant, the configuration could be removed.

Reply:
I tried removing the autoconfiguration but from what I see the `AbstractSpringComponentScanServer` is trying to call a default constructor that the `ObservationProvider` or `ObservationClientProvider` just don't have. I wonder if it makes sense to have providers that have a non default constructor :shrug:

Reply:
I will modify the component scanning logic to allow trying to autowire by constructor. That way if you have a provider that has constructors Spring will try to autowire using it

Reply:
OK I gave it another thought and I don't see any difference between `MicrometerMetricsProvider` which is defined as a Spring bean and `ObservationProvider` or `ObservationClientProvider`. Trying to change the existing code to allow construction injection is not enough for all test to pass. I would leave the current auto-configuration cause that prepares the providers the way they should be set up given that the ObservationRegistry is on the classpath. From my perspective this PR is ready.

Reply:
>  Trying to change the existing code to allow construction injection is not enough for all test to pass.

Thanks @marcingrzejszczak , do you mind if I push this specific change? Thank you.

Reply:
Oh be my guest :) 

What I did is I modified the `JAXRSServerFactoryBeanDefinitionParser` to have its last static method to fallback to constructor autowiring.

```java
static List<Object> createBeansFromDiscoveredClasses(ApplicationContext context,
                                                         Collection<Class<?>> classes,
                                                         Class<? extends Annotation> serviceClassAnnotation) {
        AutowireCapableBeanFactory beanFactory = context.getAutowireCapableBeanFactory();
        final List< Object > providers = new ArrayList<>();
        for (final Class< ? > clazz: classes) {
            if (serviceClassAnnotation != null && clazz.getAnnotation(serviceClassAnnotation) == null) {
                continue;
            }
            Object bean;
            try {
                bean = beanFactory.createBean(clazz, AutowireCapableBeanFactory.AUTOWIRE_BY_TYPE, true);
            } catch (Exception ex) {
                String stackTrace = ExceptionUtils.getStackTrace(ex);
                LOG.fine("Autowire failure for a " + clazz.getName() + " bean: " + stackTrace);
                try {
                    bean = beanFactory.createBean(clazz);
                } catch (Exception ex2) {
                    // Fallback to constructor based injection
                    String stackTrace2 = ExceptionUtils.getStackTrace(ex2);
                    LOG.fine("Bean creation failure for a " + clazz.getName() + " bean: " + stackTrace2);
                    GenericBeanDefinition definition = new GenericBeanDefinition();
                    definition.setBeanClass(clazz);
                    definition.setAutowireMode(AbstractBeanDefinition.AUTOWIRE_CONSTRUCTOR);
                    String name = clazz.getName() + "_" + System.currentTimeMillis();
                    ((BeanDefinitionRegistry) context).registerBeanDefinition(name, definition);
                    bean = context.getBean(name);
                }
            }
            providers.add(bean);
        }
        return providers;
    }
```

Reply:
@marcingrzejszczak pushed my changes
  - the `MicrometerObservationAutoConfiguration` is removed (the features are created manually now but I will take care of it)
  - refactored a bit test cases for systests/tracing (to separate Brave and Micrometer)
  - added more test cases for Spring Boot (and here was confused, asking for your help)

Hope you have no objections, thank you.

Reply:
> added more test cases for Spring Boot (and here was confused, asking for your help)

I've replied to the comment 

> Hope you have no objections, thank you.

Oh absolutely not, I'm just the guest here :) Do all the changes that you think make sense!